File: week2/assign2.js
//#region  Task1
function func1(name){ 
    // 建立各 point 座標
    let points = [
        {name: "辛巴", x: -3, y: 3, side: "L"},
        {name: "貝吉塔", x: -4, y: -1, side: "L"},
        {name: "悟空", x: 0, y: 0, side: "L"},
        {name: "特南克斯", x: 1, y: -2, side: "L"},
        {name: "丁滿", x: -1, y: 4, side: "R"},
        {name: "弗利沙", x: 4, y: -1, side: "R"},
    ];
    // 在陣列中依序尋找第一個符合角色名稱的 Object
    const target = points.find(p => p.name === name);
    // 沒找到就 return
    if (target === undefined) {
        return console.log("抱歉，無此角色。");
    }

    let shortestPoint = []; // 最短距離角色array
    let longestPoint = []; // 最長距離角色array
    let shortestDistance = 10000000; // 預設初始最短距
    let longestDistance = -1; // 預設初始最長距
    // 遍歷每個 Object
    for (let i=0; i < points.length; i++) {

        // 先排除 target 本身
        if (target.name != points[i].name) {
            // 計算兩點距離
            let distance = Math.abs(target.x - points[i].x) + Math.abs(target.y - points[i].y);
            if (target.side != points[i].side) {
                distance +=2;
            }

            // 比較最短距離
            if (distance < shortestDistance) { // 若 distance 小於至今最短距，則汰換"最短距"跟"最短point"
                shortestDistance = distance;
                // 清空 shortestPoint
                shortestPoint.length = 0; // shortestPoint=[]; 是把 shortestPoint 指到新陣列，舊陣列會保留
                shortestPoint.push(points[i].name);
            } else if (distance == shortestDistance) { // 若 distance 等於至今最短距，將 point 加入"最短point"
                shortestPoint.push(points[i].name);
            }

            // 比較最長距離
            if (distance > longestDistance) { // 若 distance 大於至今最長距，則汰換"最長距"跟"最長point"
                longestDistance = distance;
                // 清空 longestPoint
                longestPoint.length = 0;
                longestPoint.push(points[i].name);
            } else if (distance == longestDistance) { // 若 distance 等於至今最長距，將 point 加入"最長point"
                longestPoint.push(points[i].name);
            }
        }
    }

    return console.log(`最遠${longestPoint.join("、")};最近${shortestPoint.join("、")}`);
} 

func1("辛巴");  // print 最遠弗利沙；最近丁滿、⾙吉塔
 
func1("悟空");  // print 最遠丁滿、弗利沙；最近特南克斯
 
func1("弗利沙");  // print 最遠⾟巴，最近特南克斯
 
func1("特南克斯");  // print 最遠丁滿，最近悟空

//#endregion

console.log("======================================")

//#region Task2
let bookings = []; //已預約的時段 [{start: xx, end: xx, service: xx}]

function func2(ss, start, end, criteria){ 
    // 用正則表達式分開 criteria，"key"、"<= >= ="、"value"
    // arrMatch.match()回傳規則 -> arrMatch[0]=整個匹配的字串、arrMatch[1]=第一個()裡的部分1、arrMatch[1]=第二個()裡的部分...
    const match = criteria.match(/^\s*(r|c|name)\s*(>=|<=|=)\s*(.+)\s*$/);
    if (!match) 
        return console.log("條件式不符合規則。");
    const [, key, operations, value] = match; // 指派各項常數
    
    // 查詢特定服務是否可預約
    let isBooking = (name) => {
        // 用 .filter 篩選指定 service
        let serviceBookings = bookings.filter(b => b.service == name);
        // 遍歷每個已預約時段
        for (let b of serviceBookings) {
            // 判斷時段是否重疊
            if (start < b.end && end > b.start)
                return false;
        }
        return true;
    }

    // 開始判斷條件
    if (key === "name") {
        // 偵測指定 service 是否有空缺時間
        if (isBooking(value)) { 
            // 紀錄預約時間及服務項目
            bookings.push({start: start, end: end, service: value});
            return console.log(value); // 回傳結果
        }
    }
    else if (key === "r" || key === "c") {
        // 用 .filter 依照條件篩選可行的服務
        let availableServices = ss.filter(s => {
            if (operations === ">=") return s[key] >= value;  // Object[key] -> 會找 key 常數內的值來對應 Object 中的屬性
            else if (operations === "<=") return s[key] <= value;
        })

        // 用 .sort 排序，與條件 value 越靠近排越前面
        availableServices.sort((a, b) => Math.abs(a[key] - value) - Math.abs(b[key] - value));

        // 遍歷可行的服務
        for (let s of availableServices) {
            // 偵測對應 service 是否有空缺時間
            if (isBooking(s.name)) {
                // 有就紀錄並回傳
                bookings.push({start: start, end: end, service: s.name});
                return console.log(s.name);
            }
        }
    }

    return console.log("sorry");
} 

const services=[ 
{"name":"S1", "r":4.5, "c":1000}, 
{"name":"S2", "r":3, "c":1200}, 
{"name":"S3", "r":3.8, "c":800} 
];

func2(services, 15, 17, "c>=800");  // S3 
func2(services, 11, 13, "r<=4");  // S3 
func2(services, 10, 12, "name=S3");  // Sorry 
func2(services, 15, 18, "r>=4.5");  // S1 
func2(services, 16, 18, "r>=4");  // Sorry 
func2(services, 13, 17, "name=S1");  // Sorry 
func2(services, 8, 9, "c<=1500");  // S2 

//#endregion

console.log("======================================")

//#region Task3
function func3(index){ 
    if (Number.isInteger(index) === false || index < 0) 
        return console.log("不符合格式。");
    let sequence = [25];  // 初始序列
    let number = 25; // 初始數字

    // 依照規律計算序列，直到過了 index 值
    for(let i=0; i<index; i+=4) { 
        // 序列的規律是 -2, -3, +1, +2 一個循環
        number -= 2;
        sequence.push(number);
        number -= 3;
        sequence.push(number);
        number += 1;
        sequence.push(number);
        number += 2;
        sequence.push(number);
    }

    return console.log(sequence[index]);
} 
func3(1);  // print 23 
func3(5);  // print 21 
func3(10);  // print 16 
func3(30);  // print 6
//#endregion

console.log("======================================")

//#region Task4
function func4(sp, stat, n){
    const statArray = [...stat]; // 用展開運算子將 stat 拆成陣列
    let bestSub;  // 紀錄最佳"差值"的容器，用來比較"差值"
    let fitCarIndex;  // 紀錄最合適車廂編號

    // 遍歷所有車廂
    for (let i=0; i<statArray.length; i++) {
        // 篩選可用車廂
        if (statArray[i] == "0") {
            let sub = sp[i] - n;   // 計算"車廂可乘坐人數"與"乘客數"的差值

            // 如果 bestSub 還是預設值，直接汰換初始值
            if (bestSub === undefined) {
                fitCarIndex = i;
                bestSub = sub;
            }

            // 計算最合適車廂
            if (sub === 0) { // 一定是最合適的，直接輸出結果 index
                return console.log(i);
            }else if (sub > 0) { 
                // 塞的下的車廂(>0)優先於塞不下的車廂(<0) or 越接近乘車人數的優先
                if (bestSub < 0 || sub < bestSub) {
                    fitCarIndex = i;
                    bestSub = sub;
                }
            }else if (sub < 0) {
                // 越接近乘車人數的優先
                if(bestSub < sub) {
                    fitCarIndex = i;
                    bestSub = sub;
                }
            }
        }
    }
    return console.log(fitCarIndex);
    
} 
func4([3, 1, 5, 4, 3, 2], "101000", 2);  // print 5 
func4([1, 0, 5, 1, 3], "10100", 4);  // print 4 
func4([4, 6, 5, 8], "1000", 4);  // print 2 

//#endregion

File: week2/assign2.py
import re
#region Task1
def func1(name): 
    points = [{"name": "辛巴", "x": -3, "y": 3, "side": "L"},
        {"name": "貝吉塔", "x": -4, "y": -1, "side": "L"},
        {"name": "悟空", "x": 0, "y": 0, "side": "L"},
        {"name": "特南克斯", "x": 1, "y": -2, "side": "L"},
        {"name": "丁滿", "x": -1, "y": 4, "side": "R"},
        {"name": "弗利沙", "x": 4, "y": -1, "side": "R"}]
    
    # 指派 target 的資料
    target = None
    for point in points:  # 用name找對應字典
        if point["name"] == name:
            target = point
            break
    if target is None:
        return print("抱歉，無此角色。")
    
    # 初始最短、最長 distance 與 point
    shortestPoint = []
    longestPoint = []
    shortestDist = 10000
    longestDist = -1
    # 遍歷 points 算最大最小距離
    for point in points:
        # 排除 target 本身
        if point != target:
            # 計算距離
            distance = abs(point["x"] - target["x"]) + abs(point["y"] - target["y"])
            if point["side"] != target["side"]:
                distance += 2

            # 比較最短距離
            if distance < shortestDist:
                shortestPoint.clear() # 清空 list
                shortestPoint.append(point["name"])
                shortestDist = distance
            elif distance == shortestDist:
                shortestPoint.append(point["name"])

            # 比較最長距離
            if distance > longestDist:
                longestPoint.clear() # 清空 list
                longestPoint.append(point["name"])
                longestDist = distance
            elif distance == longestDist:
                longestPoint.append(point["name"])
    
    # python 中 .join() 是 string 的 methods
    return print("最遠" + "、".join(longestPoint) + ";最近" + "、".join(shortestPoint))


func1("辛巴")  # print 最遠弗利沙；最近丁滿、⾙吉塔
func1("悟空")  # print 最遠丁滿、弗利沙；最近特南克斯
func1("弗利沙")  # print 最遠⾟巴，最近特南克斯
func1("特南克斯")  # print 最遠丁滿，最近悟空

#endregion

print("=====================================")

#region Task2
bookings = [] # 已預約的時段 [{start: xx, end: xx, service: xx}]
def func2(ss, start, end, criteria): 
    # 用正則表達式區分criteria中的 key、operations、value
    match = re.match(r"^\s*(r|c|name)\s*(>=|<=|=)\s*(.+)\s*$", criteria)
    if match:
        key, operations, value = match.groups()
        # python 不會自動轉型，所以要先轉成數字型態，已利之後比大小
        if key in ("r", "c"):
            value = float(value) if "." in value else int(value) # 三元運算子: A(true) if 條件 else B(false)
    else: 
        return print("條件式不符合規則。")
    
    # 查詢特定服務是否有時段可預約
    def canBooking(service):
        # 篩選特定服務的已預約時段
        serviceBooking = [b for b in bookings if b["service"] == service] # 列表生成式 -> 可以下條件篩選element並建立new list
        for b in serviceBooking:
            # 判斷時段是否重疊
            if b["start"] < end and b["end"] > start:
                return False
        return True
    
    # 開始判斷條件
    if key == "name":
        # 檢查時段是否可預約
        if canBooking(value):
            # 紀錄本次預約時間及服務
            bookings.append({"start": start, "end": end, "service": value})
            return print(value)          
    elif key in ("r", "c"):
        # 篩選符合條件的服務(分<=、>=)
        if operations == "<=":
            availableServices = [s for s in ss if s[key] <= value]
        elif operations == ">=":
            availableServices = [s for s in ss if s[key] >= value]

        # 依照"各服務數值"與"條件數值"接近程度排序 .sort(key= 函式)
        availableServices.sort(key= lambda s: abs(s[key] - value))

        # 檢查"符合條件的服務"是否有時段可預約
        for s in availableServices:
            if canBooking(s["name"]):
                # 紀錄本次預約時間及服務
                bookings.append({"start": start, "end": end, "service": s["name"]})
                return print(s["name"])
    
    return print("sorry")

services=[ 
{"name":"S1", "r":4.5, "c":1000}, 
{"name":"S2", "r":3, "c":1200}, 
{"name":"S3", "r":3.8, "c":800} 
] 
func2(services, 15, 17, "c>=800")  # S3 
func2(services, 11, 13, "r<=4")  # S3 
func2(services, 10, 12, "name=S3")  # Sorry 
func2(services, 15, 18, "r>=4.5")  # S1 
func2(services, 16, 18, "r>=4")  # Sorry 
func2(services, 13, 17, "name=S1")  # Sorry 
func2(services, 8, 9, "c<=1500")  # S2
#endregion

print("=====================================")

#region Task3
def func3(index): 
    if type(index) != int or index <= 0:
        return print("不符合格式。")
    number = 25 # 初始數字
    patterns = [-2, -3, +1, +2] # 序列規律

    # 首項 + (一次循環總和 * 幾次循環) + (剩下不足循環的部分)
    element = number + sum(patterns) * (index // 4) + sum(patterns[0: index % 4])

    return print(element);

func3(1)  # print 23 
func3(5)  # print 21 
func3(10)  # print 16 
func3(30)  # print 6
#endregion

print("=====================================")

#region Task4
def func4(sp, stat, n): 
    bestSpaces = None  # 紀錄"最佳"車廂位置數
    bestSpacesIndex = None  # 紀錄"最佳"車廂索引
    for i, s in enumerate(stat): # 同時拿取索引值(i)跟元素(s)
        # 篩選可用車廂
        if s == "0":
            # 如果 bestSpaces 還是預設值，直接汰換初始值
            if bestSpaces is None:
                bestSpaces = sp[i]
                bestSpacesIndex = i
            
            if sp[i] == n: # 最符合，直接輸出結果 index
                return print(i)
            elif sp[i] < n: # 座位數比乘客數少
                # 且至今"最佳座位數"還少於"此車廂座位數"，調換優先集
                if bestSpaces < sp[i]: 
                    bestSpaces = sp[i]
                    bestSpacesIndex = i
            elif sp[i] > n: # 座位數比乘客數多
                # 至今"最佳座位數"少於"乘客數"或多於"此車廂座位數"，調換優先集
                if bestSpaces < n or bestSpaces > sp[i]:
                    bestSpaces = sp[i]
                    bestSpacesIndex = i
    
    return print(bestSpacesIndex)


# your code here 
func4([3, 1, 5, 4, 3, 2], "101000", 2)  # print 5 
func4([1, 0, 5, 1, 3], "10100", 4)  # print 4 
func4([4, 6, 5, 8], "1000", 4)  # print 2 
#endregion




